### ISSUE TITLE
Support C#/VB scripts

### ISSUE THREAD

**Issue by johannesegger** [ashmind/mirrorsharp#84]
[`CSharpLanguage`](https://github.com/ashmind/mirrorsharp/blob/master/Common/Internal/Roslyn/CSharpLanguage.cs) only supports full documents (i.e. you have to define a class), because it uses [`RoslynSession`](https://github.com/ashmind/mirrorsharp/blob/master/Common/Internal/Roslyn/RoslynSession.cs) which operates on a solution. It would be nice if we could have a language or a session that works with scripts (i.e. you can define methods without explicitely defining a class). What would be the preferred or the easiest way we could do this? As far as I can tell an external library can't define a new language because all of the interfaces used are marked as `internal`. I mean we could define them in a [friend assembly](https://github.com/ashmind/mirrorsharp/blob/master/Common/Properties/AssemblyInfo.cs) but that feels wrong.

**Comment by johannesegger** [comment]
Ohh, that one was easier than I thought:

```csharp
options.CSharp.ParseOptions = options.CSharp.ParseOptions.WithKind(SourceCodeKind.Script);
```

**Comment by johannesegger** [comment]
Still, Roslyn supports a special compilation for scripts where you have more options, e.g. you can specify a _globalsType_. When running the script you can pass an instance of this type which is available from the compiled script. I think we would need a special language that creates another session to handle this case, right?

**Comment by ashmind** [comment]
It's a good question. I'll have to look into it (can't promise specific time, but hope for this weekend).
My feeling is some kind of `Compilation` factory in settings. I don't think it's a separate language, but we'll see once I try to implement it.

The reason why everything is `internal` is I didn't feel the API was stable enough to expose it. But given it may take years to become "stable enough" at this pace, I'll reconsider that as well.

**Comment by ashmind** [comment]
This should be available in MirrorSharp 0.10.0 as soon as builds pass and NuGet updates.

I wasn't able to implement this using a generic enough API (because there is no easy way to change `IsSubmission` on a project) -- so it now has a specific API. You can either set mode on the application start using `SetupCSharp(o => o.SetScriptMode(hostObjectType: ...))`, or later by using `session.Roslyn.SetScriptMode(...)`.

Note that I don't provide a built-in way to change this from client, however Owin.Demo includes an demo option extension that can do that.
